pkg/spdk/backing_image.go

```diff
@@ -0,0 +1,713 @@
+	query := parsedURL.Query()
+	query.Set("forV2Creation", "true")
+	parsedURL.RawQuery = query.Encode()
+
+	size, err := httpHandler.GetSizeFromURL(parsedURL.String())
+	if err != nil {
+		return errors.Wrapf(err, "failed to get the file size from %v", fromAddress)
+	}
+	if size != int64(bi.Size) {
+		return errors.Wrapf(err, "download file %v size %v is not the same as the backing image size %v", fromAddress, size, bi.Size)
+	}
+
+	if _, err := httpHandler.DownloadFromURL(bi.ctx, parsedURL.String(), targetFh, bi); err != nil {
+		return errors.Wrapf(err, "failed to download the file from %v", fromAddress)
+	}
+	return nil
+}
+
+func (bi *BackingImage) prepareFromSync(targetFh *os.File, fromAddress, srcLvsUUID string) (err error) {
+	if fromAddress == "" || srcLvsUUID == "" {
+		return errors.Wrapf(err, "missing required source backing image service address %v or source lvsUUID %v", fromAddress, srcLvsUUID)
+	}
+	srcBackingImageServiceCli, err := GetServiceClient(fromAddress)
+	if err != nil {
+		return errors.Wrapf(err, "failed to init the source backing image spdk service client")
+	}
+	exposedSnapshotLvolAddress, err := srcBackingImageServiceCli.BackingImageExpose(bi.Name, srcLvsUUID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to expose the source backing image %v", bi.Name)
+	}
+	externalSnapshotLvolName := GetBackingImageSnapLvolName(bi.Name, srcLvsUUID)
+
+	srcIP, srcPort, err := splitHostPort(exposedSnapshotLvolAddress)
+	if err != nil {
+		return errors.Wrapf(err, "failed to split host and port from address %v", exposedSnapshotLvolAddress)
+	}
+	_, _, err = connectNVMeTarget(srcIP, srcPort, maxNumRetries, retryInterval)
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect to NVMe target for source backing image %v in lvsUUID %v with address %v", bi.Name, srcLvsUUID, exposedSnapshotLvolAddress)
+	}
+
+	bi.log.Info("Creating NVMe initiator for source backing image %v", bi.Name)
+	initiator, err := nvme.NewInitiator(externalSnapshotLvolName, helpertypes.GetNQN(externalSnapshotLvolName), nvme.HostProc)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create NVMe initiator for source backing image %v in lvsUUID %v with address %v", bi.Name, srcLvsUUID, exposedSnapshotLvolAddress)
+	}
+	if _, err := initiator.Start(srcIP, strconv.Itoa(int(srcPort)), true); err != nil {
+		return errors.Wrapf(err, "failed to start NVMe initiator for source backing image %v in lvsUUID %v with address %v", bi.Name, srcLvsUUID, exposedSnapshotLvolAddress)
+	}
+
+	bi.log.Infof("Opening NVMe device %v", initiator.Endpoint)
+	srcFh, err := os.OpenFile(initiator.Endpoint, os.O_RDWR, 0666)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open NVMe device %v for source backing image %v in lvsUUID %v with address %v", initiator.Endpoint, bi.Name, srcLvsUUID, exposedSnapshotLvolAddress)
+	}
+
+	ctx, cancel := context.WithCancel(bi.ctx)
+	defer cancel()
+	_, err = util.IdleTimeoutCopy(ctx, cancel, srcFh, targetFh, bi, true)
```

Comment:
Is `IdleTimeoutCopy` a blocking call?

Reply:
Yes.

Reply:
The whole `prepareBackingImageSnapshot` is in another go routing
the BackingImageCreate() returns already.
So we don't need to make this another go routine.